### PR TITLE
Fix Azure CLI authentication for federated token service principals

### PR DESCRIPTION
**Pull request by rauchy**
## What changes are proposed in this pull request?

**WHAT**: This PR enhances the Azure CLI authentication logic in the Databricks Go SDK to properly handle federated token service principals used in AKS workload identity scenarios.

The specific changes include:
- Enhanced MSI detection logic in `config/auth_azure_cli.go` to recognize service principals with GUID-like names as federated token authentication
- Added `isGuidLike()` helper function to detect GUID patterns (8-4-4-4-12 character format)
- Updated the authentication flow to skip tenant ID parameters for federated token service principals (treating them like MSI)
- Added comprehensive test coverage in `config/auth_azure_cli_federated_token_test.go`

**WHY**: The existing MSI detection logic only recognized system/user assigned identities by their specific names (`systemAssignedIdentity` or `userAssignedIdentity`). However, when using AKS with workload identity, service principals authenticate using federated tokens and show their client ID as the name (e.g., `5817e630-86b3-4f67-a38e-a63e6a1a401c`).

This caused the SDK to incorrectly treat federated token service principals as regular service principals, leading to:
1. SDK passing `--tenant <tenant_id>` to `az account get-access-token`
2. Azure CLI rejecting the request because federated tokens don't work with explicit tenant parameters
3. Complete authentication failure in AKS environments

The decision to use GUID pattern detection was made because:
- Federated token service principals consistently show client IDs (GUIDs) as their names
- This approach is more efficient than a fallback mechanism (no retry needed)
- It matches the authentication flow observed in working environments where no tenant parameter is used from the start
- It preserves backward compatibility for all existing authentication methods

## How is this tested?

**Unit Tests:**
- Added `TestAzureCliCredentials_FederatedTokenServicePrincipal` which simulates the exact federated token scenario using the client ID pattern from the reported issue
- Test uses `FAIL_IF_TENANT_ID_SET=true` environment variable to ensure the fix correctly skips tenant ID usage (test would fail if `--tenant` parameter was passed)
- All existing Azure CLI authentication tests continue to pass, ensuring no regressions

**Test Coverage Validation:**
- Federated token service principals: Correctly detected and skip tenant ID ✅
- Traditional MSI (system/user assigned identities): Behavior unchanged ✅
- Regular service principals: Continue to use tenant ID as before ✅
- Edge cases: GUID detection handles malformed strings appropriately ✅

The test uses the existing mock infrastructure (`testdata/az`) rather than custom mocks, ensuring consistency with other authentication tests.

### CODE REVIEW

File: config/auth_azure_cli.go [hunk 1]
```diff
@@ -176,8 +176,17 @@ func (ts *azureCliTokenSource) getTokenBytes() ([]byte, error) {
...
 		return nil, fmt.Errorf("cannot unmarshal account info: %w", err)
 	}
+
+	// Check for authentication types that should not use tenant ID:
+	// 1. Traditional MSI (system/user assigned identities)
 	isMsi := account.User.Type == "servicePrincipal" && (account.User.Name == "systemAssignedIdentity" || account.User.Name == "userAssignedIdentity")
-	if !isMsi {
+
+	// 2. Service principal with client ID as name (likely federated token)
+	// When user.name is a GUID and user.type is servicePrincipal, it's often federated token auth
```
[the remark this answers is not in the document]
Comment (rauchy) [reply config/auth_azure_cli.go:185]: Are you talking about regular service principals that use GUID client IDs but sign in with a client secret (not federated tokens)? Since both look like GUIDs, this logic can’t really tell them apart.

If that’s the issue, I can change the flow: first try with `--tenant` (same as today), and if that fails for service principals, fall back to trying without it. WDYT?
